cve-text-analyzer/main.go

```diff
@@ -0,0 +1,357 @@
+			continue
+		}
+		out = append(out, match)
+	}
+	return out
+}
+
+func printMemoryUsage() {
+	memStats := runtime.MemStats{}
+	runtime.ReadMemStats(&memStats)
+	Log("used memory:", humanize.Bytes(memStats.Alloc))
+}
+
+func flags() (args []string) {
+	flag.StringVar(&indexFile, "indexes", indexFile, "index file")
+	flag.BoolVar(&reIndex, "reindex", reIndex, "force reindexing")
+	flag.StringVar(&nvdDbFile, "nvd-db", nvdDbFile, "nvd db file")
+	flag.StringVar(&cveListDir, "cve-list-dir", cveListDir, "cve5 list data dir")
+	flag.StringVar(&cveCsvFile, "identify-cve-csv", cveCsvFile, "csv with a list of cves to identify")
+	flag.StringVar(&format, "format", format, "output format (full, csv, override)")
+	flag.IntVar(&confidenceCutoff, "confidence-cutoff", confidenceCutoff, "threshold to drop matches below confidence level")
+	flag.Parse()
+
+	// expand easier to use paths
+	indexFile = Get(homedir.Expand(indexFile))
+	nvdDbFile = Get(homedir.Expand(nvdDbFile))
+	cveListDir = Get(homedir.Expand(cveListDir))
+	cveCsvFile = Get(homedir.Expand(cveCsvFile))
+
+	return flag.Args()
+}
+
+func getIndexes() (indexes search.Indexes) {
+	f, err := os.Open(indexFile)
+
+	if !reIndex && err == nil {
+		// read prebuilt indexes
+		indexes = readIndexes(f)
+	} else {
+		// otherwise, build indexes from source data
+		indexes = createIndexes()
+		f = Get(os.OpenFile(indexFile, os.O_RDWR|os.O_TRUNC|os.O_CREATE, os.ModePerm))
+		PanicOnError(indexes.Encode(f))
+	}
+	return indexes
+}
+
+func createIndexes() search.Indexes {
+	fetchData()
+
+	indexes := processInputsToIndexes(sync.NewExecutor(-1), nvdDbFile, cveListDir)
+	Log("process took:", timer.Lap())
+
+	return indexes
+}
+
+func readIndexes(f *os.File) (indexes search.Indexes) {
+	PanicOnError(indexes.Decode(f))
+	return indexes
+}
+
+func repoDir(repo string) string {
+	return path.Join(tmpDir, strings.TrimSuffix(path.Base(repo), ".git"))
+}
+
+func fetchData() {
+	PanicOnError(os.MkdirAll(tmpDir, os.ModePerm))
+
+	executor := sync.NewExecutor(-1)
+
+	executor.Execute(func() {
+		getLatestGrypeDbResults()
+		Log("cloning grype-db took:", timer.Lap())
+	})
+
+	executor.Execute(func() {
+		getLatestCveList()
+		Log("cloning cve list took:", timer.Lap())
+	})
+
+	executor.Wait()
+}
+
+func getLatestCveList() {
+	clone(cve5ListRepo)
+}
+
+func getLatestGrypeDbResults() {
+	clone(grypeDbRepo)
+	run(repoDir(grypeDbRepo), "make download-all-provider-cache")
+}
+
+func clone(repo string) {
+	if _, err := os.ReadDir(repoDir(repo)); err == nil {
+		run(repoDir(repo), "git pull")
+	} else {
+		run(tmpDir, "git clone --depth=1 "+repo)
+	}
+}
+
+func run(dir, command string) {
+	parts := strings.Split(command, " ")
+	cmd := exec.Command(parts[0], parts[1:]...)
+	cmd.Dir = dir
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	PanicOnError(cmd.Run())
+}
+
+func cveFilePath(cve string) string {
+	parts := strings.Split(cve, "-")
+	year := parts[1]
+	num := parts[2]
+	subdir := fmt.Sprintf("%sxxx", num[0:len(num)-3])
```

Comment:
I actually got rid of this additional subpath in the data as I found it annoying, so it's just year and then the json file for the overrides.

Reply:
So like https://github.com/anchore/nvd-data-overrides/blob/main/data/2024/CVE-2024-0199.json